### PR TITLE
[#11564] Linting should be skipped for
generated files during local build
and test

### LINKED ISSUE #11564: Linting should be skipped for generated files during local build and test
This is a very minor improvement which would improve local testing efficiency in my opinion.

**Steps to reproduce**

Run `npm lint` (specifically, it affects `lint:css` and `lint:space`).

Basically, the auto generated files of `npm run build` has been linted, resulting a large number of errors thrown out. Also, json files generated by lnp tests also fails the test.

**Expected behaviour**

Only linting whatever to be pushed to the git repository, skipping all the auto-generated items (such as those listed in `.gitignore`).

**Additional info**

For the `lint:css`, issue is resolved by adding `--ignore-pattern \"src/web/dist/*.css\"` behind the current command line strings. This instructs the `stylelint` not to check files in `dist` folder (the generated ones for production).

For the `lint:space`, `lintspaces` program does not have a good way to skip certain files. Probably a better matching rules needed to match files for linting.

### PR DESCRIPTION
**Pull request by tenebrius1**
Fixes #11564 

**Outline of Solution**
Used [husky](https://www.npmjs.com/package/husky) along with [lint-staged](https://www.npmjs.com/package/lint-staged) to create a git-commit hook which runs linting on developers local machine before every commit

Add `--cache` flag to `ng lint` and `stylelint` command which generates a [`.eslintcache`](https://cn.eslint.org/docs/1.10.3/user-guide/command-line-interface#--cache) and `.stylelintcache` file so that subsequent runs only lints changed files

Used `git diff` and `grep` in Ubuntu test environment to only lint files that have changed for `lintspaces`.

### CODE REVIEW

File: package.json
Comment (sivayogasubramanian) [inline package.json:34]: Would it be better if we added this in `devDependencies` instead?

**Comment by tenebrius1** [comment]
There are some limitations to my latest commit as the `git-diff-lint` bash script does not work on Windows. I am looking into other alternatives that are cross platforms.

**Comment by tenebrius1** [comment]
I have been looking for alternative solutions the past few days and these are the few solutions that I have come across:

1. One of the ways we could do this without any dependencies is to run bash commands inside the linting scripts as suggested in [this article](https://seeliang.medium.com/how-to-update-lint-without-lint-the-whole-project-841150aa59b0). e.g. `ng lint --eslint-config static-analysis/teammates-eslint.yml $(git diff --name-only --diff-filter=ACMRTUXB origin/master | grep -E \"(.js$|.ts$|.tsx$)\")`. This solution is however still reliant on using a linux environment which is not ideal. 
    - There is a possibility that we could use [findstr](https://superuser.com/questions/300815/grep-equivalent-for-windows-7) as a replacement for grep for Windows users but that would require additional checks and probably adding dependencies like [run-script-os](https://www.npmjs.com/package/run-script-os) to run specific scripts depending on the OS.
2. One of the more [promising solution](https://duncanlew.medium.com/getting-started-with-husky-and-lint-staged-for-pre-commit-hooks-c2764d8c9ae) is to use [husky](https://www.npmjs.com/package/husky) and [lint-staged](https://github.com/okonet/lint-staged) to create a pre-commit hook which runs the linting before every `git commit`. This solution is **cross platform** but completely changes the current workflow, where linting checks are ran locally before devs can even commit to git, instead of relying on github actions to run all the linting scripts. 
3. From my observation, a large portion of the linting time actually comes from the first `ng lint` script. As such, I think that it is possible to use my original solution (using globby package to enable negated path regex) but adding the `--cache` flag to the `ng lint` command which significantly speeds it up. This solution allows for both cross-platform solution and flexibility to add in more files to be ignored in the future (if needed).

Do let me know your opinions on the proposed solutions and if anyone has seen a better way to do the `git diff` approach do let me know so I can take a look at it!

**Comment by fsgmhoward** [comment]
@tenebrius1 I think either one is fine. Regarding your concerns:
1. Yes. It would be better if cross-platform can be achieved and no external programs are added as dependency (e.g. grep)
2. It is not a concern for a change of workflow. In fact, all developers are supposed to run lint before they commit (manually). If it is added as a commit hook, it would make the process even easier and prevent people from forgetting it until they see the red cross on GitHub.
3. This sounds good also. Please decide on your own preference.

If any other one from the team has opinion on this please add on.

File: .github/workflows/component.yml
Comment (fsgmhoward) [inline .github/workflows/component.yml:41]: This should still be preserved as husky may not be running properly at the client side (or being purposely skipped). GitHub workflow for checking it again to ensure repository consistency is still necessary. "Never Trust a Client"

You need to find a way to lint the diff at workflow, or if not possible, lint everything as before.

**Review by fsgmhoward — CHANGES_REQUESTED** [review]
Great job. Just a comment for you to improve it.

File: .github/workflows/component.yml
Comment (tenebrius1) [reply .github/workflows/component.yml:41]: Thanks for the suggestion! I was also quite worried about husky not running properly on current developers machine as I think that they will have to run `npm install` again for husky to work properly. I think that the husky package should eventually be propagated to all current and new devs as they run `npm install` in their own branches, either when they download new dependencies or when they first set up the repository.

As for now, I am currently looking at using the first solution that I proposed as a way to lint the diffs at workflow, as per what you suggested.

**Comment by tenebrius1** [comment]
I was trying my best to implement a solution where I could use `git diff` and `findstr` for windows as I have previously proposed but I have reached a roadblock where I can't use the inline command (`$()`) in cmd like I can in a unix environment. I have opted to just leave the windows `lintspaces` command as it is but have modified the Ubuntu test environment's lint-spaces to only run on changed files.

**Review by fsgmhoward — APPROVED** [review]
LGTM. Thanks for the great work.

Regarding your reply in my previous comment: frontend dependencies sometimes do change, similar to the API definitions. So for devs they are supposed to make sure they are up to date before they do development (i.e. run npm install and generateTypes after they pull the repo). It is not an issue for the add in of husky. Anyway they will realize something is missing when they failed to create commits.

**Comment by NicolasCwy** [comment]
@tenebrius1 So every commit made has to pass the lint check?

@wkurniawan07 What are your thoughts on this? I get that every commit should be lint compliant but this seems like it may slow down development, especially if we squash the PRs after review. My take is that as long as it is lint compliant when we merge it, it should be ok.

**Comment by tenebrius1** [comment]
> @tenebrius1 So every commit a dev makes locally has to pass the lint check?

Yes, but it will only lint files that have changed from the master branch so it only takes a few seconds when I ran it (with the exception of the first time they commit after the change since the `.eslintcache` is not created yet).

**Comment by NicolasCwy** [comment]
@tenebrius1 I understand that the linting would only be done on the changes from the previous commit. The only problem I have which may not be agreed by everyone is that every commit you make is lint compliant, forcing developers to either fix their commits to comply (which may slow down development) or make fewer and bigger commits to get around this.

I get that the "right" thing to do would be to have every commit comply to our linting rules, but in reality I think we are more concerned about the code complying just before we merge it into the codebase.

I have roped in @wkurniawan07, a much more senior maintainer, to see what he thinks

**Comment by samuelfangjw** [comment]
Firstly, thanks for the effort taken to try to streamline the developer experience. However, I don't see any compelling reason to use a pre-commit hook for our project at this point in time.

A pre-commit hook does not guarantee that the committed files have passed our lint checks. We will still have to enforce CI checks before merging to our trunk. From the project's perspective, it provides us minimal benefit in this context.

On the flip side, it adds an extra layer of inconvenience for developers, which encourages bad habits. It's also an additional potential point of failure and adds an extra unnecessary dependency on husky, which goes against our ["God's" of the few tools we use principle](https://teammates.github.io/teammates/overview.html#principles).

Lastly, I would say this isn't really in scope for this PR. Perhaps we can continue the discussion on pre-commit hooks in the discussions pages if needed. If we come to an agreement we can open an issue and corresponding PR for this.

**Comment by tenebrius1** [comment]
Appreciate the comments and letting me know the concerns from a senior dev perspective! I did not think about how much it would affect the development process and how much benefit it would actually bring us and have removed the pre-commit hook based on the comments given.

**Comment by fsgmhoward** [comment]
Sure. So no pre-commit hooks now and it still lints the diff. This still fix the problem described in the issue.

**Comment by fsgmhoward** [comment]
I am fine with the changes. A second review is requested. @NicolasCwy @samuelfangjw @sivayogasubramanian

**Comment by zhaojj2209** [comment]
@NicolasCwy @samuelfangjw @sivayogasubramanian reminder for final review!

**Comment by tenebrius1** [comment]
Hello @NicolasCwy  @samuelfangjw  @sivayogasubramanian, would appreciate any reviews for this PR since there hasn't been any input for quite a while!

**Comment by damithc** [comment]
@NicolasCwy @samuelfangjw please help to move this PR forward ...

**Comment by damithc** [comment]
> @NicolasCwy @samuelfangjw please help to move this PR forward ...

@TEAMMATES/active-mentors this PR needs reviews ...

**Review by zhaojj2209 — APPROVED** [review]
LGTM

**Pull request merged**
